File: python/decode_image_stream.py
import time
import threading
import queue

import image_codec_types
import symbol_codec
import transform_utils
import image_decoder
import image_decode_worker

class App:
    def __init__(self, output_file, symbol_type, dim, part_num, mp, transform):
        self.output_file = output_file
        self.image_decode_worker = image_decode_worker.ImageDecodeWorker(symbol_type, dim)
        self.part_num = part_num
        self.transform = transform
        self.mp = mp

        self.calibration = image_decoder.Calibration()

        self.frame_q = queue.Queue(maxsize=4)
        self.part_q = queue.Queue(maxsize=4)
        self.fetch_image_thread = None
        self.decode_image_threads = []
        self.save_part_thread = None
        self.transform_lock = threading.Lock()
        self.running = [False]
        self.running_lock = threading.Lock()

    def IsRunning(self):
        with self.running_lock:
            return self.running[0]

    def start(self):
        with self.running_lock:
            self.running[0] = True

        def get_transform_fn():
            with self.transform_lock:
                return self.transform.clone()

        self.fetch_image_thread = threading.Thread(target=self.image_decode_worker.fetch_image_worker, args=(self.running, self.running_lock, self.frame_q, 25))
        self.fetch_image_thread.start()
        for i in range(self.mp):
            t = threading.Thread(target=self.image_decode_worker.decode_image_worker, args=(self.part_q, self.frame_q, get_transform_fn, self.calibration))
            t.start()
            self.decode_image_threads.append(t)

        def save_part_progress_cb(save_part_progress):
            s = '{} frames processed, {}/{} parts transferred, fps={:.2f}, done_fps={:.2f}, bps={:.0f}, left_time={:0>2d}d{:0>2d}h{:0>2d}m{:0>2d}s'.format(save_part_progress.frame_num, save_part_progress.done_part_num, save_part_progress.part_num, save_part_progress.fps, save_part_progress.done_fps, save_part_progress.bps, save_part_progress.left_days, save_part_progress.left_hours, save_part_progress.left_minutes, save_part_progress.left_seconds)
            print(s)

        def save_part_complete_cb():
            print('transfer done')

        def save_part_error_cb(msg):
            print('error')
            print(msg)

        self.save_part_thread = threading.Thread(target=self.image_decode_worker.save_part_worker, args=(self.running, self.running_lock, self.part_q, self.output_file, self.part_num, save_part_progress_cb, None, save_part_complete_cb, save_part_error_cb, None, None, None, None))
        self.save_part_thread.start()

    def stop(self):
        with self.running_lock:
            self.running[0] = False
        self.fetch_image_thread.join()
        self.fetch_image_thread = None
        for i in range(self.mp):
            self.frame_q.put(None)
        self.frame_q.join()
        for t in self.decode_image_threads:
            t.join()
        self.decode_image_threads.clear()
        self.part_q.put(None)
        self.part_q.join()
        self.save_part_thread.join()
        self.save_part_thread = None

if __name__ == '__main__':
    import argparse

    parser = argparse.ArgumentParser()
    parser.add_argument('output_file', help='output file')
    parser.add_argument('symbol_type', help='symbol type')
    parser.add_argument('dim', help='dim as tile_x_num,tile_y_num,tile_x_size,tile_y_size')
    parser.add_argument('part_num', type=int, help='part num')
    parser.add_argument('--mp', type=int, default=1, help='multiprocessing')
    transform_utils.add_transform_arguments(parser)
    args = parser.parse_args()

    symbol_type = symbol_codec.parse_symbol_type(args.symbol_type)
    dim = image_codec_types.parse_dim(args.dim)
    transform = transform_utils.get_transform(args)

    app = App(args.output_file, symbol_type, dim, args.part_num, args.mp, transform)
    print('start')
    app.start()
    try:
        while app.IsRunning():
            time.sleep(1)
    except KeyboardInterrupt:
        print('stop')
    app.stop()

File: src/decode_image_stream/decode_image_stream.cpp
#include <iostream>
#include <exception>
#include <thread>
#include <atomic>
#include <chrono>
#include <filesystem>

#include <boost/program_options.hpp>

#include "image_codec.h"

class App {
public:
    App(const std::string& output_file, SymbolType symbol_type, const Dim& dim, uint32_t part_num, int mp, const Transform& transform) : m_output_file(output_file), m_image_decode_worker(symbol_type, dim), m_part_num(part_num), m_mp(mp), m_transform(transform) {
    }

    bool IsRunning() { return m_running; }

    void Start() {
        m_running = true;
        auto get_transform_fn = [this] {
            std::lock_guard<std::mutex> lock(m_transform_mtx);
            return m_transform;
        };
        m_fetch_image_thread = std::make_unique<std::thread>(&ImageDecodeWorker::FetchImageWorker, &m_image_decode_worker, std::ref(m_running), std::ref(m_frame_q), 25);
        for (int i = 0; i < m_mp; ++i) {
            m_decode_image_threads.emplace_back(&ImageDecodeWorker::DecodeImageWorker, &m_image_decode_worker, std::ref(m_part_q), std::ref(m_frame_q), get_transform_fn, m_calibration);
        }
        auto save_part_progress_cb = [](const ImageDecodeWorker::SavePartProgress& save_part_progress){
            std::cout << save_part_progress.frame_num << " frames processed, " << save_part_progress.done_part_num << "/" << save_part_progress.part_num << " parts transferred, fps=" << std::fixed << std::setprecision(2) << save_part_progress.fps << ", done_fps=" << std::fixed << std::setprecision(2) << save_part_progress.done_fps << ", bps=" << std::fixed << std::setprecision(0) << save_part_progress.bps << ", left_time=" << std::setfill('0') << std::setw(2) << save_part_progress.left_days << "d" << std::setw(2) << save_part_progress.left_hours << "h" << std::setw(2) << save_part_progress.left_minutes << "m" << std::setw(2) << save_part_progress.left_seconds << "s" << std::setfill(' ') << "\n";
        };
        auto save_part_complete_cb = []() {
            std::cout << "transfer done\n";
        };
        auto save_part_error_cb = [](const std::string& msg) {
            std::cout << "error\n";
            std::cout << msg << "\n";
        };
        m_save_part_thread = std::make_unique<std::thread>(&ImageDecodeWorker::SavePartWorker, &m_image_decode_worker, std::ref(m_running), std::ref(m_part_q), m_output_file, m_part_num, save_part_progress_cb, nullptr, save_part_complete_cb, save_part_error_cb, nullptr, nullptr, nullptr, nullptr);
    }

    void Stop() {
        m_running = false;
        m_fetch_image_thread->join();
        m_fetch_image_thread.reset();
        for (size_t i = 0; i < m_mp; ++i) {
            m_frame_q.PushNull();
        }
        for (auto& t : m_decode_image_threads) {
            t.join();
        }
        m_decode_image_threads.clear();
        m_part_q.PushNull();
        m_save_part_thread->join();
        m_save_part_thread.reset();
    }

private:
    std::string m_output_file;
    ImageDecodeWorker m_image_decode_worker;
    uint32_t m_part_num = 0;
    int m_mp = 0;
    Transform m_transform;
    Calibration m_calibration;
    ThreadSafeQueue<std::pair<uint64_t, cv::Mat>> m_frame_q{128};
    ThreadSafeQueue<DecodeResult> m_part_q{128};
    std::unique_ptr<std::thread> m_fetch_image_thread;
    std::vector<std::thread> m_decode_image_threads;
    std::unique_ptr<std::thread> m_save_part_thread;
    std::mutex m_transform_mtx;
    std::atomic<bool> m_running = false;
};

int main(int argc, char** argv) {
    try {
        std::string output_file;
        std::string symbol_type_str;
        std::string dim_str;
        uint32_t part_num = 0;
        int mp = 1;
        boost::program_options::options_description desc("usage");
        auto desc_handler = desc.add_options();
        desc_handler("help", "help message");
        desc_handler("output_file", boost::program_options::value<std::string>(&output_file), "output file");
        desc_handler("symbol_type", boost::program_options::value<std::string>(&symbol_type_str), "symbol type");
        desc_handler("dim", boost::program_options::value<std::string>(&dim_str), "dim as tile_x_num,tile_y_num,tile_x_size,tile_y_size");
        desc_handler("part_num", boost::program_options::value<uint32_t>(&part_num), "part num");
        desc_handler("mp", boost::program_options::value<int>(&mp), "multiprocessing");
        add_transform_options(desc_handler);
        boost::program_options::positional_options_description p_desc;
        p_desc.add("output_file", 1);
        p_desc.add("symbol_type", 1);
        p_desc.add("dim", 1);
        p_desc.add("part_num", 1);
        boost::program_options::variables_map vm;
        boost::program_options::store(boost::program_options::command_line_parser(argc, argv).options(desc).positional(p_desc).run(), vm);
        boost::program_options::notify(vm);

        if (vm.count("help")) {
            std::cout << desc << "\n";
            return 1;
        }

        if (!vm.count("output_file")) {
            throw std::invalid_argument("output_file not specified");
        }

        if (!vm.count("symbol_type")) {
            throw std::invalid_argument("symbol_type not specified");
        }

        if (!vm.count("dim")) {
            throw std::invalid_argument("dim not specified");
        }

        if (!vm.count("part_num")) {
            throw std::invalid_argument("part_num not specified");
        }

        auto symbol_type = parse_symbol_type(symbol_type_str);
        auto dim = parse_dim(vm["dim"].as<std::string>());
        Transform transform = get_transform(vm);
        App app(output_file, symbol_type, dim, part_num, mp, transform);
        std::cout << "start\n";
        app.Start();
        while (app.IsRunning()) {
            std::this_thread::sleep_for(std::chrono::seconds(1));
            if (std::filesystem::is_regular_file("decode_image_stream.stop")) {
                std::cout << "stop\n";
                break;
            }
        }
        app.Stop();
    }
    catch (std::exception& e) {
        std::cerr << e.what() << "\n";
    }

    return 0;
}
